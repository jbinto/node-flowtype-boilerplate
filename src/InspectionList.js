export default class InspectionList {
  constructor() {
    this.inspections = {}
  }

  add(inspection) {
    console.log('---', inspection, '-----------------------------')
    const {
      // ACTION,
      // AMOUNT_FINED,
      // COURT_OUTCOME,
      ESTABLISHMENT_ADDRESS,
      ESTABLISHMENT_NAME,
      ESTABLISHMENT_STATUS,
      INFRACTION_DETAILS,
      INSPECTION_DATE,
      INSPECTION_ID,
      SEVERITY,
    } = inspection

    // File is in random order; store using a naturally sortable key
    const hashKey = INSPECTION_DATE + INSPECTION_ID

    // Group by inspection ID. Everything is the same except:
    //  (INFRACTION_*, SEVERITY)
    if (this.inspections[hashKey] == null) {
      const summary = {
        name: ESTABLISHMENT_NAME,
        address: ESTABLISHMENT_ADDRESS,
        status: ESTABLISHMENT_STATUS,
        date: INSPECTION_DATE,
        infractions: [],
      }
      this.inspections[hashKey] = summary
    }

    const infraction = {
      details: INFRACTION_DETAILS,
      severity: SEVERITY,
      // action: ACTION,
      // courtOutcome: COURT_OUTCOME,
      // amountFined: AMOUNT_FINED,
    }
    this.inspections[hashKey].infractions.push(infraction)
  }

  count() {
    return Object.keys(this.inspections).length
  }

  toObject() {
    const reverseSortedObject = input => {
      const output = {}
      Object.keys(input)
        .sort()
        .reverse()
        .forEach(key => {
          output[key] = input[key]
        })
      return output
    }

    return reverseSortedObject(this.inspections)
  }

  toArray() {
    return Object.values(this.toObject())
  }
}
