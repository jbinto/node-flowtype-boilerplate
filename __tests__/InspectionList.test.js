/** @flow */

import InspectionList from '../src/InspectionList'

const INSPECTION_1 = {
  ROW_ID: '6497',
  ESTABLISHMENT_ID: '9009514',
  INSPECTION_ID: '104037228',
  ESTABLISHMENT_NAME: 'MANDARIN',
  ESTABLISHMENTTYPE: 'Restaurant',
  ESTABLISHMENT_ADDRESS: '2200 YONGE ST ',
  ESTABLISHMENT_STATUS: 'Conditional Pass',
  MINIMUM_INSPECTIONS_PERYEAR: '3',
  INFRACTION_DETAILS: 'Operator fail to provide adequate pest control',
  INSPECTION_DATE: '2017-08-22',
  SEVERITY: 'S - Significant',
  ACTION: 'Notice to Comply',
  COURT_OUTCOME: ' ',
  AMOUNT_FINED: ' ',
}
const INSPECTION_2 = {
  ROW_ID: '6415',
  ESTABLISHMENT_ID: '9009431',
  INSPECTION_ID: '104009693',
  ESTABLISHMENT_NAME: 'MAIN KITCHEN - BASEMENT',
  ESTABLISHMENTTYPE: 'Serving Kitchen',
  ESTABLISHMENT_ADDRESS: '301 FRONT ST W',
  ESTABLISHMENT_STATUS: 'Conditional Pass',
  MINIMUM_INSPECTIONS_PERYEAR: '3',
  INFRACTION_DETAILS: 'Operator fail to properly maintain mechanical washer',
  INSPECTION_DATE: '2017-07-12',
  SEVERITY: 'S - Significant',
  ACTION: 'Notice to Comply',
  COURT_OUTCOME: ' ',
  AMOUNT_FINED: ' ',
}
const INSPECTION_1_2ND_VIOLATION = {
  ROW_ID: '6496',
  ESTABLISHMENT_ID: '9009514',
  INSPECTION_ID: '104037228',
  ESTABLISHMENT_NAME: 'MANDARIN',
  ESTABLISHMENTTYPE: 'Restaurant',
  ESTABLISHMENT_ADDRESS: '2200 YONGE ST ',
  ESTABLISHMENT_STATUS: 'Conditional Pass',
  MINIMUM_INSPECTIONS_PERYEAR: '3',
  INFRACTION_DETAILS: 'Operator fail to properly maintain rooms',
  INSPECTION_DATE: '2017-08-22',
  SEVERITY: 'M - Minor',
  ACTION: 'Notice to Comply',
  COURT_OUTCOME: ' ',
  AMOUNT_FINED: ' ',
}

describe('xxx', () => {
  let list
  beforeEach(() => {
    list = new InspectionList()
  })

  it('adds a inspection to the list', () => {
    list.add(INSPECTION_1)
    expect(list.count()).toEqual(1)
    expect(list).toMatchSnapshot()
  })

  it('adds 2 inspections from different establishments to the list', () => {
    list.add(INSPECTION_1)
    list.add(INSPECTION_2)
    expect(list.count()).toEqual(2)
    expect(list).toMatchSnapshot()
  })

  it('groups multiple violations in the same inspection', () => {
    list.add(INSPECTION_1)
    list.add(INSPECTION_1_2ND_VIOLATION)
    expect(list.count()).toEqual(1)
    expect(list).toMatchSnapshot()
  })
})
