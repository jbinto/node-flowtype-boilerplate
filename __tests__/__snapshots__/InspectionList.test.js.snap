// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InspectionList .add() 1`] = `
InspectionList {
  "inspections": Object {
    "2017-08-22104037228": Object {
      "address": "2200 YONGE ST ",
      "date": "2017-08-22",
      "infractions": Array [
        Object {
          "details": "Operator fail to provide adequate pest control",
          "severity": "S - Significant",
        },
      ],
      "name": "MANDARIN",
      "status": "Conditional Pass",
    },
  },
}
`;

exports[`InspectionList .add() 2 inspections from different establishments 1`] = `
InspectionList {
  "inspections": Object {
    "2017-07-12104009693": Object {
      "address": "301 FRONT ST W",
      "date": "2017-07-12",
      "infractions": Array [
        Object {
          "details": "Operator fail to properly maintain mechanical washer",
          "severity": "S - Significant",
        },
      ],
      "name": "MAIN KITCHEN - BASEMENT",
      "status": "Conditional Pass",
    },
    "2017-08-22104037228": Object {
      "address": "2200 YONGE ST ",
      "date": "2017-08-22",
      "infractions": Array [
        Object {
          "details": "Operator fail to provide adequate pest control",
          "severity": "S - Significant",
        },
      ],
      "name": "MANDARIN",
      "status": "Conditional Pass",
    },
  },
}
`;

exports[`InspectionList .add() groups multiple violations in the same inspection 1`] = `
InspectionList {
  "inspections": Object {
    "2017-08-22104037228": Object {
      "address": "2200 YONGE ST ",
      "date": "2017-08-22",
      "infractions": Array [
        Object {
          "details": "Operator fail to provide adequate pest control",
          "severity": "S - Significant",
        },
        Object {
          "details": "Operator fail to properly maintain rooms",
          "severity": "M - Minor",
        },
      ],
      "name": "MANDARIN",
      "status": "Conditional Pass",
    },
  },
}
`;

exports[`InspectionList .toArray() 1`] = `
Array [
  Object {
    "address": "2200 YONGE ST ",
    "date": "2017-08-22",
    "infractions": Array [
      Object {
        "details": "Operator fail to provide adequate pest control",
        "severity": "S - Significant",
      },
      Object {
        "details": "Operator fail to properly maintain rooms",
        "severity": "M - Minor",
      },
    ],
    "name": "MANDARIN",
    "status": "Conditional Pass",
  },
  Object {
    "address": "301 FRONT ST W",
    "date": "2017-07-12",
    "infractions": Array [
      Object {
        "details": "Operator fail to properly maintain mechanical washer",
        "severity": "S - Significant",
      },
    ],
    "name": "MAIN KITCHEN - BASEMENT",
    "status": "Conditional Pass",
  },
]
`;

exports[`InspectionList .toObject() 1`] = `
Object {
  "2017-07-12104009693": Object {
    "address": "301 FRONT ST W",
    "date": "2017-07-12",
    "infractions": Array [
      Object {
        "details": "Operator fail to properly maintain mechanical washer",
        "severity": "S - Significant",
      },
    ],
    "name": "MAIN KITCHEN - BASEMENT",
    "status": "Conditional Pass",
  },
  "2017-08-22104037228": Object {
    "address": "2200 YONGE ST ",
    "date": "2017-08-22",
    "infractions": Array [
      Object {
        "details": "Operator fail to provide adequate pest control",
        "severity": "S - Significant",
      },
      Object {
        "details": "Operator fail to properly maintain rooms",
        "severity": "M - Minor",
      },
    ],
    "name": "MANDARIN",
    "status": "Conditional Pass",
  },
}
`;
